Python 3: Port three files in fetch/

stephenmcgruer opened the pull request.

stephenmcgruer committed:
Python 3: Port three files in fetch/

Hexcles commented on fetch/api/resources/infinite-slow-response.py:
```diff
@@ -12,24 +12,25 @@ def stash_write(request, key, value):
 
     # Writing an initial 2k so browsers realise it's there. *shrug*
-    response.writer.write("." * 2048)
+    response.writer.write(u"." * 2048)
```
```suggestion
    response.writer.write(b"." * 2048)
```
semantically more correct here

Hexcles commented on fetch/api/resources/infinite-slow-response.py:
```diff
@@ -12,24 +12,25 @@ def stash_write(request, key, value):
         if abortKey and request.server.stash.take(abortKey, url_dir(request)):
             break
-        response.writer.write(".")
+        if not response.writer.write(u"."):
```
nit
```suggestion
        if not response.writer.write(b"."):
```

stephenmcgruer commented:
@Hexcles - this is possibly-incomplete (haven't actually properly ported the handlers according to the guide, just made them not broken), but this appears to work on py2 and py3 for fetch/api/abort/general.any.js at least.

We could probably drop the flush altogether.

(fyi @ziransun )

Hexcles reviewed (commented):
The fix itself is sane. However, I'd like to suggest dropping the `flush` method on `ResponseWriter` altogether to prevent any future confusion.

Digging through stdlib, I don't think there's any need to call `flush` ever. In both Python 2 and 3, the base HTTP server inherits from `SocketServer.StreamRequestHandler` with the default `wbufsize = 0`.
* In Python 2, `wfile.write()` calls `flush()` automatically if `wbufsize` is 0: https://github.com/python/cpython/blob/8d21aa21f2cbc6d50aab3f420bb23be1d081dac4/Lib/socket.py#L328
* In Python 3, `wfile` in fact does not have a buffer at all and `flush()` is a no-op: https://github.com/python/cpython/blob/47a2955589bdb1a114d271496ff803ad73f954b8/Lib/socketserver.py#L791

In both versions, `wfile.write()` ultimately calls `sock.sendall` which either succeeds sending all data or raises an exception: https://docs.python.org/3/library/socket.html#socket.socket.sendall

So basically, the reliable way to detect aliveness of a socket is to check `wfile.write()`. `flush()` is either not necessary (in Python 2) or non-functional (in Python 3), and should be removed for clarity.

stephenmcgruer committed:
Text -> bytes

stephenmcgruer committed:
fixups

stephenmcgruer commented:
> However, I'd like to suggest dropping the flush method on ResponseWriter altogether to prevent any future confusion.

Sgtm, but I suggest we do that as a follow up. There aren't many call-sites, but it will muddy the water to do it here too.

I have updated the files with what I think the porting guide asks for. I'm not sure about `request.url_parts.path.split`, I think that requires that the split argument has the same type as the string, so maybe I should be isomorphic-ing it?

Hexcles commented on fetch/api/resources/stash-take.py:
```diff
@@ -4,6 +4,6 @@
 @json_handler
 def main(request, response):
     dir = '/'.join(request.url_parts.path.split('/')[:-1]) + '/'
```
I think this should be
```suggestion
    dir = u'/'.join(request.url_parts.path.split(u'/')[:-1]) + u'/'
```

Basically `request.url_parts.path` is unfortunately still a native string (#23362 ), and mixing it with Unicode literals works everywhere (in Python 2 you can mix `unicode` with `str` even though they are different, and in Python 3 they are the same). Yeah I know...

Hexcles reviewed (approved):
LGTM

Make sure to highlight the change in wptserve in the commit subject/message.

stephenmcgruer committed:
url_dir
